### PR TITLE
ENGKNOW-1792-GOR-support-pattern-based-SorceType-matching

### PR DESCRIPTION
**Pull request by gmagnu**

### CODE REVIEW

File: drivers/src/main/java/org/gorpipe/oci/driver/OCIUrl.java [hunk 1]
```diff
@@ -107,10 +109,6 @@ private static OCIUrl parseHttpUrl(URI uri) throws MalformedURLException {
 
     //  {@literal https://<oci-endpoint>/n/<namespace>/b/<bucket>/o/<path>/<to>/<file>}
     private static OCIUrl parseOCIHttpUrl(URI uri) throws MalformedURLException {
-        if (!Arrays.asList("http", "https", "s3").contains(uri.getScheme())) {
-            throw new MalformedURLException("S3 url must have http[s]/s3 scheme: " + uri);
-        }
-
         var result = new OCIUrl();
 
         result.endpoint = uri.getAuthority();
```
Comment (gumson) [inline drivers/src/main/java/org/gorpipe/oci/driver/OCIUrl.java:114]: A lot of hardcoded indices, are we sure that this is always the same?

File: drivers/src/main/java/org/gorpipe/oci/driver/OCIUrl.java [hunk 2]
```diff
@@ -156,8 +154,8 @@ private static OCIUrl parseS3HttpUrl(URI uri) throws MalformedURLException {
     }
 
     private static OCIUrl parseGlobalUrl(URI uri) throws MalformedURLException {
```
Comment (gumson) [inline drivers/src/main/java/org/gorpipe/oci/driver/OCIUrl.java:156]: See that we are using the URI object. Wasnt there a performance hit using it?

File: model/src/main/java/org/gorpipe/gor/driver/PluggableGorDriver.java [hunk 3]
```diff
@@ -45,15 +45,19 @@
...
 import static org.gorpipe.gor.driver.meta.DataType.LINK;
 
 public class PluggableGorDriver implements GorDriver {
     private final static Logger log = LoggerFactory.getLogger(PluggableGorDriver.class);
 
-    private final TreeMap<String, SourceType> protocolToSourceType = new TreeMap<>();
-    private final Map<SourceType, SourceProvider> sourceTypeToSourceProvider = new HashMap<>();
+    private final TreeMap<SourceType, SourceProvider> sourceTypeToSourceProvider = new TreeMap<>(new Comparator<SourceType>() {
+        @Override
+        public int compare(SourceType o1, SourceType o2) {
```
Comment (gumson) [inline model/src/main/java/org/gorpipe/gor/driver/PluggableGorDriver.java:56]: Is the priority used to distinguish between general http urls and OCI http urls?

File: drivers/src/main/java/org/gorpipe/oci/driver/OCIUrl.java [hunk 4]
```diff
@@ -82,22 +92,14 @@ public static OCIUrl parse(URI uri) throws MalformedURLException {
...
     private static OCIUrl parseHttpUrl(URI uri) throws MalformedURLException {
-        if (uri.getAuthority() == null) {
-            throw new MalformedURLException("Expected authority in OCI objectstore url: " + uri);
-        }
-
-        if (!uri.getAuthority().contains(".objectstorage.")) {
-            throw new MalformedURLException("Invalid OCI objectstore url: " + uri);
-        }
-
         if (uri.getPath().contains("/n/")) {
```
[the remark this answers is not in the document]
Comment (gmagnu) [reply drivers/src/main/java/org/gorpipe/oci/driver/OCIUrl.java:103]: The native OCI url is of the form https://<oci-endpoint>/n/<namespace>/b/<bucket>/o/<path>/<to>/<file> we could maybe do some more pattern matching but that is also more expensive.

File: drivers/src/main/java/org/gorpipe/oci/driver/OCIUrl.java (hunk 1 above)
Comment (gmagnu) [reply drivers/src/main/java/org/gorpipe/oci/driver/OCIUrl.java:114]: Which one are you talking about.  The structure of the urls follows a pattern, so many of them should be fixed.

File: drivers/src/main/java/org/gorpipe/oci/driver/OCIUrl.java (hunk 2 above)
Comment (gmagnu) [reply drivers/src/main/java/org/gorpipe/oci/driver/OCIUrl.java:156]: It was in the dictionaries when you 1M uris to parse, and you are only using small portion to actually open files.  Here, compared to open the file the overhead should be acceptable.

Comment (gmagnu) [reply drivers/src/main/java/org/gorpipe/oci/driver/OCIUrl.java:156]: Plus, here we need full parsing of the urls (as we are extracting so much info from it), which we did not really need in the dicts.

File: model/src/main/java/org/gorpipe/gor/driver/meta/SourceType.java [hunk 5]
```diff
@@ -67,6 +67,11 @@ public boolean supportsPreparation() {
         return false;
     }
 
+    // Default priority is 10000, 0 is highest priority.
+    public int getPriority() {
+        return 10000;
```
[the remark this answers is not in the document]
Comment (gmagnu) [reply model/src/main/java/org/gorpipe/gor/driver/meta/SourceType.java:72]: The are kind of random, creating an enum with is a good idea.

File: model/src/main/java/org/gorpipe/gor/driver/PluggableGorDriver.java (hunk 3 above)
Comment (gmagnu) [reply model/src/main/java/org/gorpipe/gor/driver/PluggableGorDriver.java:56]: Kind of, letting the OCI get the first change of picking them up.

Comment (gmagnu) [reply model/src/main/java/org/gorpipe/gor/driver/PluggableGorDriver.java:56]: So the highest priority should be on those that are very specific, then generic ones, files (that picks up the rest) should have the lowest.

**Review by gumson — APPROVED** [review]
LGTM, one suggestion on the "/n/" test

File: drivers/src/main/java/org/gorpipe/oci/driver/OCIUrl.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (gmagnu) [reply drivers/src/main/java/org/gorpipe/oci/driver/OCIUrl.java:103]: Yes I will add that

**Pull request merged**